### PR TITLE
feat(vector store): Add usage demo and enhance Redis integration

### PR DESCRIPTION
**Pull request by PingXie**
- Add Jupyter notebook in `docs/` for using langchain vector store with Memorystore Redis.
- Update `vector_store.py` for improved performance and reliability with Redis.
- Modify `__init__.py` to reflect package structure changes and new functionalities.
- Include `requirements.txt` and `setup.py` for easy installation and dependency management.
- Add `state_of_the_union.txt` in `docs/` as a sample dataset for the notebook demo.

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/langchain-google-memorystore-redis-python/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: docs/state_of_the_union.txt
Comment (glsmith3) [inline docs/state_of_the_union.txt:1]: Why is there a State of the Union text here?

Comment (PingXie) [reply docs/state_of_the_union.txt:1]: lifted from langchain's code base - https://www.google.com/search?q=langchain+src&rlz=1CAFOSO_enUS963&oq=langchain+src&gs_lcrp=EgZjaHJvbWUyBggAEEUYOTIICAEQABgWGB4yBggCEEUYPDIGCAMQRRg8MgYIBBBFGDwyBggFEEUYQDIGCAYQRRhAMgYIBxBFGEDSAQg1Nzc2ajBqN6gCALACAA&sourceid=chrome&ie=UTF-8


But I am also uncertain whether this is appropriate. @kurtisvg thoughts?

Comment (PingXie) [reply docs/state_of_the_union.txt:1]: closed offline with @kurtisvg. we can keep this text.

File: docs/vector_store.ipynb [hunk 1]
```diff
@@ -1,79 +1,356 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Installing the LangChain Memorystore Redis Module\n",
+    "\n",
+    "Interaction with the Memorystore for Redis instance from LangChain requires installing the necessary module:"
```
Comment (glsmith3) [inline docs/vector_store.ipynb:40]: "To allow LangChain to interact with the Memorystore for Redis instance, you must install the necessary module:"

File: docs/vector_store.ipynb [hunk 2]
```diff
@@ -1,79 +1,356 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Method 1: Classmethod for Direct Insertion\n",
+    "\n",
+    "This approach combines embedding creation and insertion into a single step using the from_documents classmethod:"
```
Comment (glsmith3) [inline docs/vector_store.ipynb:140]: add commas "embedding, creation, and insertion..."

File: docs/vector_store.ipynb [hunk 3]
```diff
@@ -1,79 +1,356 @@
...
+    ")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Method 2: Instance-Based Insertion\n",
+    "This approach offers flexibility when working with a new or existing RedisVectorStore:\n",
```
Comment (glsmith3) [inline docs/vector_store.ipynb:164]: suggest --> "This approach offers flexibility when working with a new or existing RedisVectorStore Instance in LangChain:\n",

File: docs/vector_store.ipynb [hunk 4]
```diff
@@ -1,79 +1,356 @@
...
+    "\n",
+    "* [Optional] Create a RedisVectorStore Instance: Instantiate a RedisVectorStore object for customization. If you already have an instance, proceed to the next step.\n",
+    "* Add Text with Metadata: Provide raw text and metadata to the instance. Embedding generation and insertion into the vector store are handled automatically."
```
Comment (glsmith3) [inline docs/vector_store.ipynb:167]: "Embedding, generation, and insertion"

File: docs/vector_store.ipynb [hunk 5]
```diff
@@ -1,79 +1,356 @@
...
+    "    texts=[d.page_content for d in docs], metadatas=[d.metadata for d in docs]\n",
+    ")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Perfor a Similarity Search (KNN)\n",
```
Comment (glsmith3) [inline docs/vector_store.ipynb:189]: "Perform"

File: docs/vector_store.ipynb [hunk 6]
```diff
@@ -1,79 +1,356 @@
...
+    "\n",
+    "With the vector store populated, it's possible to search for text semantically similar to a query.  Here's how to use KNN (K-Nearest Neighbors) with default settings:\n",
+    "\n",
+    "* Formulate the Query: A natural language question expresses the search intent (e.g., \"What did the president say about Ketanji Brown Jackson\").\n",
+    "* Retrieve Similar Results: The `similarity_search` method finds items in the vector store closest in meaning to the query."
```
Comment (glsmith3) [inline docs/vector_store.ipynb:194]: "meaning of the query"

File: docs/vector_store.ipynb [hunk 7]
```diff
@@ -1,79 +1,356 @@
...
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Perform a Range-Based Similarity Search\n",
+    "\n",
+    "Range queries provide more control by specifying a desired similarity threshold along with the query text:\n",
+    "\n",
+    "* Formulate the Query: A natural language question defines the search intent.\n",
+    "* Set Similarity Threshold: The distance_threshold parameter determines how close a match needs to be considered relevant.\n",
```
Comment (glsmith3) [inline docs/vector_store.ipynb:220]: "how close a match must be to be considered relevant.\n"

File: docs/vector_store.ipynb [hunk 8]
```diff
@@ -1,79 +1,356 @@
...
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Perform a Maximal Marginal Relevance (MMR) Search\n",
+    "\n",
+    "MMR queries aim to find results that are both relevant to the query and diverse from each other, reducing redundancy in search results.\n",
+    "\n",
+    "* Formulate the Query: A natural language question defines the search intent.\n",
+    "* Balance Relevance and Diversity: The lambda_mult parameter controls the trade-off between strict relevance and promoting variety in the results.\n",
```
Comment (glsmith3) [inline docs/vector_store.ipynb:244]: "Balance, Relevance, and Diversity"

**Review by glsmith3 — COMMENTED** [review]
Hi Ping, I left some small comments. Otherwise LGTM

File: docs/vector_store.ipynb [hunk 9]
```diff
@@ -1,79 +1,348 @@
...
-    "nbformat_minor": 4
+ "cells": [
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Google Database\n",
+    "\n",
+    "Use [Google Memorystore for Redis](https://cloud.google.com/memorystore) as a vector store for LangChain."
```
[the remark this answers is not in the document]
Comment (PingXie) [reply docs/vector_store.ipynb:10]: done

File: docs/vector_store.ipynb [hunk 10]
```diff
@@ -1,79 +1,356 @@
...
+   "source": [
+    "## Pre-reqs"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Setting Up a Memorystore Redis Instance\n",
```
[the remark this answers is not in the document]
Comment (PingXie) [reply docs/vector_store.ipynb:25]: done

File: docs/vector_store.ipynb [hunk 11]
```diff
@@ -1,79 +1,356 @@
...
+    "* Create a Memorystore for Reids Instance (v7.2): If an instance doesn't exist, follow the instructions at https://cloud.google.com/memorystore/docs/redis/create-instance-console to create a new one. Ensure version 7.2 is selected.\n",
+    "* Obtain Endpoint: Note the endpoint associated with the instance."
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Installing the LangChain Memorystore Redis Module\n",
```
[the remark this answers is not in the document]
Comment (PingXie) [reply docs/vector_store.ipynb:38]: done

**Review by craig08 — APPROVED** [review]

File: docs/vector_store.ipynb (hunk 4 above)
Comment (PingXie) [reply docs/vector_store.ipynb:167]: well - embedding is a noun here :-) think of it as "vector" so this should read as "vector generation and insertion. "

File: docs/vector_store.ipynb (hunk 6 above)
Comment (PingXie) [reply docs/vector_store.ipynb:194]: I still feel that "to" is more appropriate. rephrased it as "closest to the query in meaning". WDYT?

File: docs/vector_store.ipynb (hunk 8 above)
Comment (PingXie) [reply docs/vector_store.ipynb:244]: Balance is a verb in this context and the tradeoff is between "relevance" and "diversity"

File: docs/vector_store.ipynb (hunk 5 above)
Comment (PingXie) [reply docs/vector_store.ipynb:189]: thanks!

File: src/langchain_google_memorystore_redis/vector_store.py [hunk 12]
```diff
@@ -331,17 +344,18 @@ def drop_index(client: redis.Redis, index_name: str, index_only: bool = False):
...
             redis.RedisError: If any Redis-specific error occurs during the operation. This
                 includes connection issues, authentication failures, or errors from executing
                 the command to drop the index. Callers should handle these exceptions to
                 manage error scenarios gracefully.
         """
-        command = (
-            f"FT.DROPINDEX {index_name} {'KEEPDOCS' if index_only else ''}".strip()
-        )
+        if index_only == False:
+            raise ValueError("Not supported")
```
[the remark this answers is not in the document]
Comment (PingXie) [reply src/langchain_google_memorystore_redis/vector_store.py:356]: our module can't support it, yet

File: src/langchain_google_memorystore_redis/vector_store.py [hunk 13]
```diff
@@ -268,14 +267,33 @@ def __init__(
...
+            )
+
+        self._client = client
         self.index_name = index_name
         self.embedding_service = embedding_service
-        self.key_prefix = key_prefix + ":" if key_prefix is not None else ""
+        self.key_prefix = self.get_key_prefix(index_name)
         self.content_field = content_field
         self.vector_field = vector_field
+        self.encoding = client.get_encoder().encoding
```
[the remark this answers is not in the document]
Comment (PingXie) [reply src/langchain_google_memorystore_redis/vector_store.py:291]: yes we can. validated :-)

File: src/langchain_google_memorystore_redis/vector_store.py [hunk 14]
```diff
@@ -268,14 +267,33 @@ def __init__(
...
+        if embedding_service == None:
+            raise ValueError(
+                "An 'embedding_service' must be provided to initialize RedisVectorStore"
+            )
+
+        self._client = client
         self.index_name = index_name
         self.embedding_service = embedding_service
-        self.key_prefix = key_prefix + ":" if key_prefix is not None else ""
+        self.key_prefix = self.get_key_prefix(index_name)
```
[the remark this answers is not in the document]
Comment (PingXie) [reply src/langchain_google_memorystore_redis/vector_store.py:288]: I think the indirection would help in the meanwhile. I can remove it when merging into main - if there is no concern with not providing key prefix customization.

File: src/langchain_google_memorystore_redis/vector_store.py [hunk 15]
```diff
@@ -268,14 +267,33 @@ def __init__(
...
         self.content_field = content_field
         self.vector_field = vector_field
+        self.encoding = client.get_encoder().encoding
+
+    @staticmethod
+    def get_key_prefix(index_name: str, key_prefix: Optional[str] = None):
+        return key_prefix + ":" if key_prefix is not None else index_name + ":"
```
[the remark this answers is not in the document]
Comment (PingXie) [reply src/langchain_google_memorystore_redis/vector_store.py:295]: I don't have a strong opinion here so yeah I can remove it and just concatenate the two

File: src/langchain_google_memorystore_redis/vector_store.py [hunk 16]
```diff
@@ -94,7 +100,7 @@ def __init__(
         type: str,
         distance_strategy: DistanceStrategy,
         vector_size: int,
-        data_type: str,
+        data_type: str = "FLOAT32",
```
[the remark this answers is not in the document]
Comment (PingXie) [reply src/langchain_google_memorystore_redis/vector_store.py:103]: done

**Pull request merged**